2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+
+    `stack` has built-in support for working with Nix, although it works
+    slightly different than `cabal`.
+
+    When using `stack` with Nix, you need to do two things.  One is to pass the
+    `--nix` flag to `stack`.  The other is to edit the `stack.yaml` file and
+    specify a location of a Nix file containing a derivation.  This derivation
+    should contain all the system libraries that are dependencies for the
+    Haskell libraries you are using.  For instance, if you are developing a
+    package that depends on the `postgresql-simple` package, the derivation
+    needs to contain the `libpq` and `zlib` system libraries.
+
+    When you run `stack --nix build`, `stack` internally executes `nix-shell`
+    with the derivation you have provided.  It then re-executes itself in this
+    new environment.  Now that it is running in an environment with all the
+    required system libraries, it starts to compile all your Haskell
+    dependencies.  Finally, it compiles your actual project.
+
+    Unlike with `cabal`, `stack` has no way of directly using a GHC package
+    database provided by the user.  Unfortunately, there is no way of building
+    Haskell packages with Nix and making `stack` use them.
+
+    This method is nice for people and teams already using `stack`, but have
+    complicated system dependencies.
+
+    The `stack` documentation has a
+    [nice section](https://docs.haskellstack.org/en/stable/nix_integration/)
+    about integration with Nix.
+
+If you're looking to use Nix for Haskell development, I recommend setting up
+your CI to use `nix-build`, and then using `cabal` + Nix for everyday
+development.  If your team is already using `stack`, then using Nix to provide
+system dependencies can be a nice way to make your project fully reproducible
+(since `stack` only deals with making the Haskell-part reproducible).
+
+### 3. How can I specify the version of Haskell packages to use?
+
+When building with `nix-build` directly, or using `cabal` + Nix, you need a way
+to specify to Nix the versions of the Haskell packages you depend on.  For
+instance, if you depend on an older version of `aeson`, you need to make sure
+that Nix doesn't accidentally use a recent version.
+
+In general, there are two ways to fix Haskell package versions.  One way is to
+use the Haskell package set defined in nixpkgs, and overwrite individual
+packages that should be different.  The other way is to completely generate your
+own Nix Haskell package set.
+
+1.  _Overwrite individual entries in the Haskell package set in nixpkgs._
+
+    As explained above, nixpkgs has a built-in Haskell package set.  In
+    general, each package is set to the latest version on Hackage[^8].
+
+    If you need to use an older version of a package, you can easily overwrite
+    packages in the nixpkgs Haskell package set.  For instance, if you need to
+    an older version of `aeson`, you can just specify it in your Nix files.
```

Comment:
So does this mean that if you want to build a package with both `cabal` and `nix` you need to do some additional work to make their configurations consistent with each other and make them use the same versions?